Alerting Settings handling and access

skrech opened the pull request:
# Description

THIS IS DEPENDENT ON https://github.com/trento-project/web/pull/3462

This PR introduces changes on how alerting configuration is made and how this configuration is later accessed. 

Up  until now, configuration of the alerting and mailer settings were only supplied by a combination of compile-time (per-build target) parameters and runtime-supplied environment variables. Now, it's possible to get, set and partially update the these settings over the a web API. The alerting infrastructure is updated to get the settings from a centralized AlertingSettings model/module and configure the mail templates, as well as mailer configuration with these settings (as opposed to directly getting them from the Application environment). To keep backwards-compatibility and to support "declarative configuration" the possibility to configure via Application environment (env-vars) is preserved, but the following precedence scheme is implemented:

1. If _any_ of the following env-vars is set, their values take precedence: `ENABLE_ALERTING`, `ALERTING_SENDER`, `ALERTING_RECIPIENT`, `SMTP_SERVER`, `SMTP_PORT`, `SMTP_USERNAME`, `SMTP_PASSWORD`. If user didn't specify some of the env-vars, trento provides default values.

2. If all of the previously specified env-vars are not set, then configuration will be made with settings get from the DB. These settings can be changed (idempotent set or partially updated) on demand via a web API. In a next PR, a web-ui will be provided.

Note: Settings returned by the GET API call would always return the values that are currently in effect (if any). Changes to the settings requested over the API would be accepted only when no env-var configuration is in place.

Adds fixes towards [TRNT-2711](https://jira.suse.com/browse/TRNT-2711)

## How was this tested?

Modified unit tests that depend on the newly added model fields `enforced_from_env`. 
Split test for alerting settings coming from the DB from the ones that check alerting configuration from the app env.

## Did you update the documentation?

Will be updated with another PR.

nelsonkopliku reviewed (changes requested):
Hey @skrech thanks for this! It's a very good baseline.

The main reason I am requesting changes is the testing aspect.
It seems to me we need to extend those.

Happy to answer any question.

nelsonkopliku commented on test/trento_web/controllers/v1/settings_controller_test.exs:
A similar thinking as in this comment https://github.com/trento-project/web/pull/3457#discussion_r2073078834 applies here as well

nelsonkopliku commented on lib/trento/settings.ex:
```diff
@@ -182,33 +192,71 @@ defmodule Trento.Settings do
   # Alerting Settings
 
+  @spec alerting_settings_enforced_from_env? :: boolean()
+  def alerting_settings_enforced_from_env? do
```
I would make it private, yes. And related specific tests removed altogether.

It would be tested anyway when we test `get_alerting_settings`, `create_alerting_settings` and `update_alerting_settings`. Covering those with the relevant cases (env vs db settings) covers also that internal function.

skrech commented on lib/trento/infrastructure/alerting/alerting.ex:
```diff
@@ -17,69 +17,79 @@ defmodule Trento.Infrastructure.Alerting.Alerting do
-
-  defp maybe_notify_critical_host_health(true, host_id) do
-    %HostReadModel{hostname: hostname} = Trento.Repo.get!(HostReadModel, host_id)
+  def notify_critical_host_health(host_id) do
```
Looks great, will implement that! :) Thanks!

skrech commented on test/support/helpers/alerting_settings_helper.ex:
```diff
@@ -0,0 +1,29 @@
+
+  import ExUnit.Callbacks
+
+  def nil_alerting_app_env do
```
The thing is that I don't want to perceive this as `initial` state. I treat the initial state as undefined and this function explicitly "nullify" the app env. My intention was to use nil as verb -- as nullify_alerting_app_env, but shorted. I might change it to `clear_alerting_app_env`?

nelsonkopliku replied:
I see, feel free to either keep as initially suggested or pick any of `clear_alerting_app_env`, `nillified_alerting_app_env` `nullified_alerting_app_env` 😄

skrech commented on test/trento/infrastructure/alerting/alerting_test.exs:
Great, I'll leave these tests as they are and thoroughly test the contexts.

nelsonkopliku reviewed (commented):
I am realizing we might be missing one last test: the actual precedence of env settings over db ones, unless I am not seeing it 👀 

What I mean is a test where we have some settings in the database, but we do have settings from the env, and what gets exposed is the one from env 🙈

skrech commented on test/trento/settings_test.exs:
```diff
@@ -949,11 +957,68 @@ defmodule Trento.SettingsTest do
+        ] do
+      @scenario scenario
+
+      test "return full settings with default values when only #{case_name} is set" do
```
Actually not a typo but apparently not right choice of words. Maybe using `complete` would do?

nelsonkopliku replied:
deal 👍

skrech commented:
> I am realizing we might be missing one last test: the actual precedence of env settings over db ones, unless I am not seeing it 👀

Indeed, I'll write a test for that, thank you.

skrech commented:
Re-requesting review because I added a settings controller test that modifies global state. Because of that, I moved that test into a separate module.

nelsonkopliku reviewed (approved):
:shipit:
